plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java

```diff
@@ -164,38 +180,39 @@ public DeleteResult delete() throws IOException {
-                ObjectListing list;
-                if (prevListing != null) {
-                    final ObjectListing finalPrevListing = prevListing;
-                    list = SocketAccess.doPrivileged(() -> clientReference.get().listNextBatchOfObjects(finalPrevListing));
-                } else {
-                    final ListObjectsRequest listObjectsRequest = new ListObjectsRequest();
-                    listObjectsRequest.setBucketName(blobStore.bucket());
-                    listObjectsRequest.setPrefix(keyPath);
-                    listObjectsRequest.setRequestMetricCollector(blobStore.listMetricCollector);
-                    list = SocketAccess.doPrivileged(() -> clientReference.get().listObjects(listObjectsRequest));
-                }
-                final List<String> blobsToDelete = new ArrayList<>();
-                list.getObjectSummaries().forEach(s3ObjectSummary -> {
+            ListObjectsV2Iterable listObjectsIterable = SocketAccess.doPrivileged(
+                () -> clientReference.get()
+                    .listObjectsV2Paginator(
+                        ListObjectsV2Request.builder()
+                            .bucket(blobStore.bucket())
+                            .prefix(keyPath)
+                            .overrideConfiguration(
+                                o -> o.addMetricPublisher(blobStore.getStatsMetricPublisher().listObjectsMetricPublisher)
+                            )
+                            .build()
+                    )
+            );
+
+            Iterator<ListObjectsV2Response> listObjectsResponseIterator = listObjectsIterable.iterator();
+            while (listObjectsResponseIterator.hasNext()) {
+                ListObjectsV2Response listObjectsResponse = SocketAccess.doPrivileged(listObjectsResponseIterator::next);
+                List<String> blobsToDelete = listObjectsResponse.contents().stream().map(s3Object -> {
                     deletedBlobs.incrementAndGet();
-                    deletedBytes.addAndGet(s3ObjectSummary.getSize());
-                    blobsToDelete.add(s3ObjectSummary.getKey());
-                });
-                if (list.isTruncated()) {
-                    doDeleteBlobs(blobsToDelete, false);
-                    prevListing = list;
-                } else {
-                    final List<String> lastBlobsToDelete = new ArrayList<>(blobsToDelete);
-                    lastBlobsToDelete.add(keyPath);
-                    doDeleteBlobs(lastBlobsToDelete, false);
-                    break;
+                    deletedBytes.addAndGet(s3Object.size());
+
+                    return s3Object.key();
+                }).collect(Collectors.toList());
+
+                if (!listObjectsResponseIterator.hasNext()) {
+                    blobsToDelete.add(keyPath);
                 }
+
+                doDeleteBlobs(blobsToDelete, false);
             }
-        } catch (final AmazonClientException e) {
+        } catch (SdkException e) {
```

Comment:
Should we be catching `RuntimeException` here to avoid bubbling up errors such as permissions? I've had this problem in discovery-ec2 where I couldn't figure out where my call was actually failing, changing it to RuntimeException (SdkException is a child of it) put the error in the logs with the right stack.

Reply:
Alright yeah that makes sense, we should catch RuntimeException for better debuggability